\documentclass[a4paper]{article}
\usepackage{url} % For \url{}
%\VignetteIndexEntry{Getting started with genoPlotR}
%\VignettePackage{genoPlotR}
\newcommand{\code}[1]{\texttt{#1}}
\newcommand{\pkg}[1]{{\normalfont\fontseries{b}\selectfont #1}}
\newcommand{\genoPlotR}{\pkg{genoPlotR}}
\newcommand{\R}{{\sffamily R}}

\title{Getting started with \genoPlotR{}}
\author{Lionel Guy}

\begin{document}

\maketitle

\tableofcontents

\section*{Introduction}

The \genoPlotR{} package is intended to produce publication-grade graphics 
of gene and genome maps. With the amazing speed of data production of new 
DNA sequencing techniques and the increase in the number of software 
available to compare these sequences, there is a great need to graphically 
represent these sequences and their comparisons. A number of packages 
already exist (Artemis, ACT, mauve), but none of them produces easily 
reproducible, publication-grade graphics. The goal of this package is 
to fill in that gap. 

This document provides an introduction to \genoPlotR{}, providing the user 
with examples of increasing complexity. It is not meant as a comprehensive
guide otto all the functions and options of the package, but rather as 
a first approach to the package.

To load the library in a R session, type:

<<results=hide>>=
library(genoPlotR)
@ 
\section{Quick start}

Loading the simplest dataset, applying a color scheme and some limits to the 
plotting aread, adding a tree and some annotations.
<<>>=
data(three_genes)
comparisons[[1]]$col <- apply_color_scheme(c(0.6, 0.4, 0.5), "grey")
names <- c("Huey", "Dewey", "Louie")
names(dna_segs) <- names
tree <- newick2phylog("(((Huey:4.2,Dewey:3.9):3.1,Louie:7.3):1);")
mid_pos <- middle(dna_segs[[1]])
xlims <- list(c(Inf, -Inf), c(-Inf, Inf), c(1850, 2800))
annot <- annotation(x1=c(mid_pos[1], dna_segs[[1]]$end[2]),
                     x2=c(NA, dna_segs[[1]]$end[3]),
                     text=c(dna_segs[[1]]$name[1], "region1"),
                     rot=c(30, 0), col=c("grey", "black"))
@

Now plotting these three segments:

\begin{center}
<<fig=TRUE, width=6, height=2.5, echo=TRUE>>=
plot_gene_map(dna_segs=dna_segs, comparisons=comparisons,
              annotations=annot, annotation_height=1.3,
              tree=tree, tree_width=2,
              xlims=xlims,
              main="Comparison of Huey, Dewey and Louie")
@ 
\end{center}

\section{Getting help}

There are various ways to get help with \genoPlotR{}. First, you can start 
the general help in a web browser, and then click on ``packages'' and 
find \genoPlotR{} in the list. It will provide a list of the functions 
available. The first link in the list leads to a very general description of
the package.
<<eval=FALSE, results=hide>>=
help.start()
@ 
Another way of obtaining the list of functions present in the package is to run 
<<eval=FALSE>>=
library(help=genoPlotR)
@ 
A lot of examples and help are available in the main functions, i.e. the 
reading functions (the various \code{read\_dna\_seg\_from}* and 
\code{read\_comparison\_from}* functions) and the main plotting 
function, \code{plot\_gene\_map}.
<<eval=FALSE>>=
help("read_functions")
help("plot_gene_map")
@ 
Finally, the web page on R-forge (\url{http://genoplotr.r-forge.r-project.org/})
provides ways to get in touch with the \genoPlotR{} community and to submit 
bugs and feature requests.

\section{Objects in \genoPlotR{}}

This section will give an overview of the different types of \R{} objects in
\genoPlotR{}.

\subsection{\code{dna\_seg}}

A \code{dna\_seg} object are is a collection of genes or elements along a 
genome, to be represented on a map. 

\code{dna\_seg} objects need to have 4 columns, \code{name}, \code{start}, 
\code{end} and \code{strand}. Extra columns with names \code{col}, 
\code{lty}, \code{lwd}, \code{pch}, \code{cex}, \code{gene\_type} will 
be used in the  plotting process. Other extra columns will be kept in 
the object, but not used.

<<>>=
names1 <- c("feat1", "feat2", "feat3")
starts1 <- c(2, 1000, 1050)
ends1 <- c(600, 800, 1345)
strands1 <- c("-", -1, 1)
cols1 <- c("blue", "grey", "red")
df1 <- data.frame(name=names1, start=starts1, end=ends1,
                  strand=strands1, col=cols1)

dna_seg1 <- dna_seg(df1)
str(dna_seg1)
@

\subsection{\code{comparison}}

A \code{comparison} is a collection of similarities, representing the
comparison between two DNA segments. 

Objects (either data frames or lists) should have at least named
elements \code{start1}, \code{end1}, \code{start2} and \code{end2}. 
In addition, it can use a \code{col}or column to give. Additional 
numeric columns can be used for color-coding (via \code{apply\_color\_scheme}.

<<>>=
starts1 <- c(2, 1000, 1050)
ends1 <- c(600, 800, 1345)
starts2 <- c(50, 800, 1200)
ends2 <- c(900, 1100, 1322)
comparison1 <- as.comparison(data.frame(start1=starts1, end1=ends1,
                                        start2=starts2, end2=ends2))
str(comparison1)
@ 

\subsection{\code{annotation}}

An \code{annotation} object is used to annotate a DNA segment. It has 
labels attached to positions. Each label can be attached to a single 
position or to a range. 

<<>>=
mid_pos <- middle(dna_segs[[1]])
annot1 <- annotation(x1=mid_pos, text=dna_segs[[1]]$name)
str(annot1)
@ 

\subsection{\code{tree}}

A tree description in Newick format can ba parsed using \pkg{ade4} package.
<<>>=
tree <- newick2phylog("(((A_aaa:4.2,B_bbb:3.9):3.1,C_ccc:7.3):1);")
str(tree$leaves)
@ 

\section{Reading data}

\subsection{DNA segments}

Several formats can be read by \genoPlotR{} to produce \code{dna\_seg} objects:
\begin{itemize}
  \item EMBL files (\code{read\_dna\_seg\_from\_embl})
  \item Genbank files (\code{read\_dna\_seg\_from\_genbank})
  \item PTT (protein table) files (basically a tabular version of a Genbank 
    file)(\code{read\_dna\_seg\_from\_ptt})
  \item Tab files (user generated) (\code{read\_dna\_seg\_from\_ptt})
\end{itemize}
The function \code{read\_dna\_seg\_from\_file} is a wrapper function, that
will attempt to guess the correct format of the file.

The first three files are common biological formats and can be downloaded 
from major databases, such as the NCBI (\url{http://www.ncbi.nlm.nih.gov/}) 
and the EMBL (\url{http://www.ebi.ac.uk/embl/Access/index.html}). The 
definition of EMBL and Genbank files can be found at 
\url{http://www.ebi.ac.uk/embl/Documentation/FT_definitions/feature_table.html}.

\subsection{Comparisons}

\genoPlotR{} can read tabular files, either user-generated tab files 
(\code{read\_comparison\_from\_tab}), or from BLAST output
(\code{read\_comparison\_from\_blast}). To produce files that are readable by 
\genoPlotR{}, the \code{-m 8} or \code{9} option should be used in 
\code{blastall}, or \code{-outfmt 6} or {7} with the BLAST+ suite.

\subsection{Mauve output}

The backbone output of the Mauve genome aligner 
(\url{http://asap.ahabs.wisc.edu/mauve/index.php}) can be parsed using 
\code{read\_comparison\_from\_blast}\footnote{Tested with Mauve 2.3.1}.

The function will return a list consisting of a list of \code{dna\_seg}s
and the corresponding \code{comparisons}.

\section{Plotting data}

There is only one plotting function in \genoPlotR{}, \code{plot\_gene\_map}.
Many arguments are available, but here is a list of the most important. Check
the documentation for a more thorough description.
\begin{itemize}
  \item[\code{dna\_segs}] A list of DNA segment objects.
  \item[\code{comparisons}] A list of comparisons. Should containt one element
    less than the previous.
  \item[\code{tree}] An eventual phylogenetic tree to be plotted at the
    left of the figure.
  \item[\code{annotations}] An annotation object, or a list of annotations. 
    Will display annotations to the first, or to all DNA segments, respectively.
  \item[\code{xlims}] A list of even-numbered numeric vectors, giving the
    borders of sub-segments to be plotted. The vector \code{c(1,5000,8000,6000}
    will display two sub-segments (1 to 5000 and 6000 to 8000), the second 
    being in reverse orientation.
  \item[\code{main}] A title to the plot.
  \item[\code{scale}] Should a scale be displayed at the bottom right of the
    plot?
  \item[\code{dna\_seg\_scale}] Allows to control the addition of scales
    to each segments. If simply \code{TRUE}, will display a scale on each
    segment. If a vector, a scale will be displayed for the corresponding 
    \code{TRUE} element.
  \item[\code{global\_color\_scheme}] Allows to recalcultate the colors of 
    the comparisons, to have colors corresponding to the same scale for all 
    comparisons.
  \item[\code{plot\_new}] Turn off to avoid creating a new plot. Escpecially
    useful to integrate a \genoPlotR{} plot in a larger figure.
\end{itemize}

\section{Other useful functions}

\begin{itemize}
  \item[\code{apply\_color\_scheme}] Allows to apply a gray scale or hues
    of red and blue to a comparison.
  \item[\code{middle}] Useful to get the middle of a gene, especially to
    create annotations.
  \item[Datasets] Type \code{data(package="genoPlotR")} to get the full list.
\end{itemize}

\section{Examples}

\subsection{A very simple example}

\subsection{Mauve alignment of four \emph{Bartonella} genomes}

\subsection{Several sub-segments of four \emph{Bartonella} genomes}

\subsection{Two segments of the Y chromsome in human and chimp}

\subsection{Generating data online}

<<echo=FALSE, results=hide>>=
@ 

\end{document}
